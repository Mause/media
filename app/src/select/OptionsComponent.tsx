import { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import useSWR from 'swr';
import { Alert, Breadcrumbs, Typography } from '@mui/material';
import { useAuth0 } from '@auth0/auth0-react';
import * as _ from 'lodash-es';

import * as qs from '../qs';
import { subscribe, getToken } from '../utils';
import type { Torrents } from '../streaming';
import { Loading } from '../render';
import type { components } from '../schema';
import { DisplayError } from '../DisplayError';
import { MonitorAddComponent } from '../MonitorComponent';
import { MLink } from '../MLink';
import { DisplayTorrent } from '../DisplayTorrent';
import type { ManualAddComponentState } from '../ManualAddComponent';

import { Shared } from './SeasonSelectComponent';

export type ITorrent = components['schemas']['ITorrent'];
type ProviderSource = components['schemas']['ProviderSource'];

const ranking = [
  'Movies/XVID',
  'Movies/x264',
  'Movies/x264/720',
  'Movies/XVID/720',
  'Movies/BD Remux',
  'Movies/Full BD',
  'Movies/x264/1080',
  'Movies/x264/4k',
  'Movies/x265/4k',
  'Movies/x264/3D',
  'Movs/x265/4k/HDR',

  'TV Episodes',
  'TV HD Episodes',
  'TV UHD Episodes',
  'TV-UHD-episodes',
];

function remove(bit: string): string {
  if (bit.startsWith('Mov')) {
    const parts = bit.split('/');
    return parts.slice(1).join('/');
  }
  return bit;
}

function OptionsComponent({ type }: { type: 'movie' | 'series' }) {
  const { season, episode, tmdb_id } = useParams();

  const { data: meta } = useSWR<{ title: string }>(
    (season ? 'tv' : 'movie') + '/' + tmdb_id,
  );
  const { data: torrents } = useSWR<Torrents>('torrents');
  const {
    items: results,
    loading,
    errors,
  } = useSubscribes<ITorrent>(
    `/api/stream/${type}/${tmdb_id}?` + qs.stringify({ season, episode }),
  );
  const dt = (result: ITorrent) => (
    <DisplayTorrent
      tmdb_id={tmdb_id!}
      season={parseInt(season!)}
      episode={parseInt(episode!)}
      torrents={torrents}
      torrent={result}
    />
  );
  const grouped = _.groupBy(results, 'category');
  const auto = _.maxBy(
    grouped['x264/1080'] || grouped['TV HD Episodes'] || [],
    'seeders',
  );
  const bits = _.sortBy(
    Object.entries(grouped),
    ([category]) => -ranking.indexOf(category),
  ).map(([category, results]) => (
    <div key={category}>
      <h3>{remove(category)}</h3>
      {_.sortBy(results, (i) => -i.seeders).map((result) => (
        <li key={result.title}>{dt(result)}</li>
      ))}
    </div>
  ));

  let header = null;
  if (type === 'movie') {
    header = (
      <Breadcrumbs aria-label="breadcrumb">
        <Shared />
        <Typography color="textPrimary">{meta?.title}</Typography>
      </Breadcrumbs>
    );
  } else {
    const url = `/select/${tmdb_id}/season`;
    header = (
      <Breadcrumbs aria-label="breadcrumb">
        <Shared />
        <MLink color="inherit" to={url}>
          {meta?.title}
        </MLink>
        <MLink color="inherit" to={`${url}/${season}`}>
          Season {season}
        </MLink>
        <Typography color="textPrimary">Episode {episode}</Typography>
      </Breadcrumbs>
    );
  }

  return (
    <div>
      {header}
      <div style={{ textAlign: 'center' }}>
        <Loading loading={loading.length !== 0} large={true} />
      </div>
      {Object.entries(errors).map(([key, error]) => (
        <DisplayError
          key={key}
          message={`Error occured whilst loading options from ${key}`}
          error={error}
        />
      ))}
      {loading.map((source) => (
        <Alert key={source} color="info">
          Loading options from {source}
        </Alert>
      ))}
      {bits.length || loading.length !== 0 ? (
        <div>
          <p>Auto selection: {auto ? dt(auto) : 'None'}</p>
          <ul>{bits}</ul>
        </div>
      ) : (
        <div>
          <p>No results</p>
        </div>
      )}
      <ul>
        <li>
          <MLink
            to="/manual"
            state={
              {
                tmdb_id: tmdb_id!,
                season,
                episode,
              } satisfies ManualAddComponentState
            }
          >
            Add manually
          </MLink>
        </li>
        <li>
          <MonitorAddComponent
            tmdb_id={parseInt(tmdb_id!)}
            type={type === 'movie' ? 'MOVIE' : 'TV'}
          />
        </li>
        <li>
          <MLink
            to={{
              pathname: `/websocket/${tmdb_id}`,
              search: qs.stringify({ season, episode }),
            }}
          >
            Search with websockets
          </MLink>
        </li>
      </ul>
    </div>
  );
}

interface SubscriptionShape<T> {
  items: T[];
  name: string;
  loading: boolean;
  error?: Error;
}

function useSubscribe<T>(
  baseUrl: string,
  name: string,
  authorization?: string,
): SubscriptionShape<T> {
  const url = baseUrl + '&source=' + name;
  const [subscription, setSubscription] = useState<SubscriptionShape<T>>({
    items: [],
    loading: true,
    name,
    error: undefined,
  });

  useEffect(() => {
    if (!authorization) return; // don't subscribe until we have auth

    const items: T[] = [];
    return subscribe<T>(
      url,
      (data) => {
        items.push(data);
        setSubscription({
          items,
          name,
          loading: true,
        });
      },
      (error) => setSubscription({ name, error, loading: false, items }),
      authorization,
      () => setSubscription({ name, loading: false, items }),
    );
  }, [url, authorization, name]);

  return subscription;
}

function useToken() {
  const auth = useAuth0();
  const [token, setToken] = useState<string>();
  useEffect(() => {
    void getToken(auth).then(setToken);
  }, [auth]);
  return token;
}

function useSubscribes<T>(url: string): {
  items: T[];
  loading: string[];
  errors: { [key: string]: Error };
} {
  const token = useToken();

  const p: ProviderSource[] = [
    /*
    'rarbg',
    'horriblesubs',
    'kickass',
    */
    'torrentscsv',
    'nyaasi',
    'piratebay',
  ];
  const providers = [
    useSubscribe<T>(url, p[0], token),
    useSubscribe<T>(url, p[1], token),
    useSubscribe<T>(url, p[2], token),
    /*
    useSubscribe<T>(url, p[3], token),
    useSubscribe<T>(url, p[4], token),
    useSubscribe<T>(url, p[5], token),
    */
  ];

  return {
    items: providers
      .map((t) => t.items || [])
      .reduce((a, b) => a.concat(b), []),
    loading: providers.filter((t) => t.loading).map((t) => t.name),
    errors: Object.fromEntries(
      providers.filter((t) => t.error).map((t, i) => [p[i], t.error]),
    ) as { [key: string]: Error },
  };
}

export { OptionsComponent };
