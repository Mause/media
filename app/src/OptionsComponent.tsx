import _ from 'lodash';
import qs from 'qs';
import React, { useState, useEffect } from 'react';
import { subscribe, MLink } from './utils';
import { Torrents } from './streaming';
import { useParams } from 'react-router-dom';
import useSWR from 'swr';
import { Loading } from './render';
import { Breadcrumbs, Typography } from '@material-ui/core';
import { Shared } from './SeasonSelectComponent';
import { DownloadState } from './DownloadComponent';
import { DisplayError } from './IndexComponent';

function getHash(magnet: string) {
  const u = new URL(magnet);
  return _.last(u.searchParams.get('xt')!.split(':'));
}

export interface ITorrent {
  source: 'KICKASS' | 'RARBG' | 'HORRIBLESUBS';
  title: string;
  seeders: number;
  download: string;
  category: string;
  episode_info: { seasonnum?: string; epnum?: string };
}

export function DisplayTorrent({
  torrent,
  torrents,
  tmdb_id,
  season,
  episode,
}: {
  season?: string;
  episode?: string;
  tmdb_id: string;
  torrent: ITorrent;
  torrents?: Torrents;
}) {
  const state: DownloadState = {
    downloads: [
      {
        tmdb_id: tmdb_id,
        magnet: torrent.download,
        season: season,
        episode: episode,
      },
    ],
  };
  const url = { pathname: '/download', state };
  return (
    <span>
      <strong title={torrent.source}>{torrent.source.substring(0, 1)}</strong>
      &nbsp;
      <MLink to={url}>{torrent.title}</MLink>
      &nbsp;
      <small>{torrent.seeders}</small>
      &nbsp;
      {torrents && getHash(torrent.download)! in torrents && (
        <small>downloaded</small>
      )}
    </span>
  );
}

const ranking = [
  'Movies/XVID',
  'Movies/x264',
  'Movies/x264/720',
  'Movies/XVID/720',
  'Movies/BD Remux',
  'Movies/Full BD',
  'Movies/x264/1080',
  'Movies/x264/4k',
  'Movies/x265/4k',
  'Movies/x264/3D',
  'Movs/x265/4k/HDR',

  'TV Episodes',
  'TV HD Episodes',
  'TV UHD Episodes',
  'TV-UHD-episodes',
];

function remove(bit: string): string {
  if (bit.startsWith('Mov')) {
    const parts = bit.split('/');
    return parts.slice(1).join('/');
  }
  return bit;
}

function OptionsComponent({ type }: { type: 'movie' | 'series' }) {
  const { season, episode, tmdb_id } = useParams<{
    tmdb_id: string;
    season?: string;
    episode?: string;
  }>();

  const { data: meta } = useSWR<{ title: string }>(
    (season ? 'tv' : 'movie') + '/' + tmdb_id,
  );
  const { data: torrents } = useSWR<Torrents>('torrents');
  const { items: results, loading, error } = useSubscribe<ITorrent>(
    `/stream/${type}/${tmdb_id}?` + qs.stringify({ season, episode }),
  );
  const dt = (result: ITorrent) => (
    <DisplayTorrent
      tmdb_id={tmdb_id}
      season={season}
      episode={episode}
      torrents={torrents}
      torrent={result}
    />
  );
  const grouped = _.groupBy(results, 'category');
  const auto = _.maxBy(
    grouped['Movies/x264/1080'] || grouped['TV HD Episodes'] || [],
    'seeders',
  );
  const bits = _.sortBy(
    _.toPairs(grouped),
    ([category]) => -ranking.indexOf(category),
  ).map(([category, results]) => (
    <div key={category}>
      <h3>{remove(category)}</h3>
      {_.sortBy(results, (i) => -i.seeders).map((result) => (
        <li key={result.title}>{dt(result)}</li>
      ))}
    </div>
  ));

  let header = null;
  if (type === 'movie') {
    header = (
      <Breadcrumbs aria-label="breadcrumb">
        <Shared />
        <Typography color="textPrimary">{meta && meta.title}</Typography>
      </Breadcrumbs>
    );
  } else {
    const url = `/select/${tmdb_id}/season`;
    header = (
      <Breadcrumbs aria-label="breadcrumb">
        <Shared />
        <MLink color="inherit" to={url}>
          {meta && meta.title}
        </MLink>
        <MLink color="inherit" to={`${url}/${season}`}>
          Season {season}
        </MLink>
        <Typography color="textPrimary">Episode {episode}</Typography>
      </Breadcrumbs>
    );
  }

  return (
    <div>
      {header}
      <div style={{ textAlign: 'center' }}>
        <Loading loading={loading} large={true} />
      </div>
      {error && <DisplayError error={error} />}
      {bits.length || loading ? (
        <div>
          <p>Auto selection: {auto ? dt(auto) : 'None'}</p>
          <ul>{bits}</ul>
        </div>
      ) : (
        <div>
          <p>No results</p>
        </div>
      )}
      <ul>
        <li>
          <MLink
            to={{
              pathname: '/manual',
              state: { tmdb_id, season, episode },
            }}
          >
            Add manually
          </MLink>
        </li>
        <li>
          <MLink
            to={{
              pathname: `/monitor/add/${tmdb_id}`,
              state: { type: type === 'movie' ? 'MOVIE' : 'TV' },
            }}
          >
            Add to monitor
          </MLink>
        </li>
      </ul>
    </div>
  );
}

function useSubscribe<T>(url: string) {
  const [subscription, setSubscription] = useState<{
    items: T[];
    loading: boolean;
    error?: Error;
  }>({ loading: true, items: [], error: undefined });

  useEffect(() => {
    const items: T[] = [];
    subscribe<T>(
      url,
      (data) => {
        items.push(data);
        setSubscription({
          items,
          loading: true,
        });
      },
      (error) => setSubscription({ error, loading: false, items }),
      () => setSubscription({ loading: false, items }),
    );
  }, [url]);

  return subscription;
}

export { OptionsComponent };
