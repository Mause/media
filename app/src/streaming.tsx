import * as Sentry from '@sentry/react';
import { ErrorInfo } from 'react';
import { Helmet } from 'react-helmet-async';
import {
  BrowserRouter as Router,
  useLocation,
  Route,
  Routes,
} from 'react-router-dom';
import { ErrorBoundary, FallbackProps } from 'react-error-boundary';
import { Grid, Link as MaterialLink } from '@mui/material';
import { styled } from '@mui/material/styles';
import { SWRConfig } from 'swr';
import { useProfiler } from '@sentry/react';
import { useAuth0 } from '@auth0/auth0-react';

import { IndexComponent } from './IndexComponent';
import {
  EpisodeSelectComponent,
  SeasonSelectComponent,
} from './SeasonSelectComponent';
import { StatsComponent } from './StatsComponent';
import { SearchComponent } from './SearchComponent';
import { OptionsComponent } from './OptionsComponent';
import { load, MLink, ExtMLink } from './utils';
import {
  MonitorComponent,
  MonitorAddComponent,
  MonitorDeleteComponent,
} from './MonitorComponent';
import { ManualAddComponent } from './ManualAddComponent';
import { DownloadComponent } from './DownloadComponent';
import { DownloadAllComponent } from './DownloadAllComponent';
import { Websocket } from './Websocket';
import { components } from './schema';
import { DiagnosticsComponent } from './DiagnosticsComponent';
import Storybook from './Storybook';

if (import.meta.env.NODE_ENV === 'production') {
  Sentry.init({
    dsn: 'https://8b67269f943a4e3793144fdc31258b46@sentry.io/1869914',
    release: import.meta.env.REACT_APP_HEROKU_SLUG_COMMIT,
    environment: 'development',
    integrations: [Sentry.browserTracingIntegration()],
    tracesSampleRate: 0.75, // must be present and non-zero
  });
}

export type TorrentFile = components['schemas']['InnerTorrentFile'];
export type Torrents = { [key: string]: components['schemas']['InnerTorrent'] };
export type IndexResponse = components['schemas']['IndexResponse'];
export type MovieResponse = components['schemas']['MovieDetailsSchema'];
export type SeriesResponse = components['schemas']['SeriesDetails'];
export type EpisodeResponse = components['schemas']['EpisodeDetailsSchema'];

function RouteTitle({
  title,
  children,
}: {
  title: string;
  children: React.ReactNode;
}) {
  return (
    <>
      <Helmet>
        <title>{title}</title>
      </Helmet>
      {children}
    </>
  );
}

function reportError(error: Error, info: ErrorInfo) {
  Sentry.withScope((scope) => {
    scope.setExtras({
      componentStack: info.componentStack,
      digest: info.digest,
    });
    const eventId = Sentry.captureException(error);
    Sentry.showReportDialog({ eventId });
  });
}

const PREFIX = 'ParentComponentInt';
const classes = {
  root: `${PREFIX}-root`,
};
const NavRoot = styled('nav')(({ theme }) => ({
  [`& .${classes.root}`]: {
    margin: theme.spacing(1),
    linkStyle: 'underline',
  },
}));

const Login = () => {
  const { loginWithRedirect, isAuthenticated, logout } = useAuth0();

  if (isAuthenticated) {
    return (
      <MaterialLink href="#" onClick={() => void logout({})} underline="hover">
        Logout
      </MaterialLink>
    );
  } else {
    return (
      <MaterialLink
        href="#"
        onClick={() => void loginWithRedirect({})}
        underline="hover"
      >
        Login
      </MaterialLink>
    );
  }
};

function ParentComponentInt() {
  useProfiler('ParentComponentInt');

  const auth = useAuth0();
  console.log({ user: auth.user });

  return (
    <Router>
      <h1>Media</h1>

      <NavRoot className={classes.root}>
        <Grid container spacing={1}>
          <Grid size={{ xs: 'auto' }}>
            <MLink to="/">Home</MLink>
          </Grid>
          <Grid size={{ xs: 'auto' }}>
            <MLink to="/monitor">Monitors</MLink>
          </Grid>
          <Grid size={{ xs: 'auto' }}>
            <ExtMLink href="http://novell.mause.me:9091">Transmission</ExtMLink>
          </Grid>
          <Grid size={{ xs: 'auto' }}>
            <ExtMLink href="https://app.plex.tv">Plex</ExtMLink>
          </Grid>
          {auth.user && <Grid size={{ xs: 'auto' }}>{auth.user.name}</Grid>}
          <Grid size={{ xs: 'auto' }}>
            <Login />
          </Grid>
        </Grid>
      </NavRoot>

      <br />

      <ErrorBoundary
        onError={reportError}
        FallbackComponent={(props: FallbackProps) => {
          const error = props.error as Error;
          return (
            <div>
              An error has occured:
              <code>
                <pre>
                  {error.message}
                  {error.stack
                    ?.toString()
                    .split('\n')
                    .map((line: string) => (
                      <span key={line}>
                        {line}
                        <br />
                      </span>
                    ))}
                </pre>
              </code>
              <button onClick={props.resetErrorBoundary}>Retry</button>
            </div>
          );
        }}
      >
        <AppRoutes />
      </ErrorBoundary>
    </Router>
  );
}
function SwrConfigWrapper({
  WrappedComponent,
}: {
  WrappedComponent: React.ComponentType;
}) {
  const auth = useAuth0();
  return (
    <SWRConfig
      value={{
        // five minute refresh
        refreshInterval: 5 * 60 * 1000,
        fetcher: async (path: string, params: string) =>
          await load(
            path,
            params,
            auth.isAuthenticated
              ? {
                  Authorization:
                    'Bearer ' + (await auth.getAccessTokenSilently()),
                }
              : {},
          ),
      }}
    >
      <WrappedComponent />
    </SWRConfig>
  );
}

export function swrConfig(WrappedComponent: React.ComponentType) {
  return () => <SwrConfigWrapper WrappedComponent={WrappedComponent} />;
}
const ParentComponent = swrConfig(ParentComponentInt);

function AppRoutes() {
  const auth = useAuth0();
  const location = useLocation();

  if (!(auth.isAuthenticated || location.pathname === '/storybook')) {
    return <div>Please login</div>;
  }

  return (
    <Routes>
      <Route
        path="*"
        element={
          <RouteTitle title="Page not Found">
            <div>Page not found</div>
          </RouteTitle>
        }
      />
      <Route
        path="/websocket/:tmdbId"
        element={
          <RouteTitle title="Websocket">
            <Websocket />
          </RouteTitle>
        }
      />
      <Route
        path="/select/:tmdb_id/options"
        element={
          <RouteTitle title="Movie Options">
            <OptionsComponent type="movie" />
          </RouteTitle>
        }
      />

      <Route
        path="/select/:tmdb_id/season/:season/episode/:episode/options"
        element={
          <RouteTitle title="TV Options">
            <OptionsComponent type="series" />
          </RouteTitle>
        }
      />
      <Route
        path="/select/:tmdb_id/season/:season/download_all"
        element={
          <RouteTitle title="Download Season">
            <DownloadAllComponent />
          </RouteTitle>
        }
      />
      <Route
        path="/select/:tmdb_id/season/:season"
        element={
          <RouteTitle title="Select Episode">
            <EpisodeSelectComponent />
          </RouteTitle>
        }
      />
      <Route
        path="/select/:tmdb_id/season"
        element={
          <RouteTitle title="Select Season">
            <SeasonSelectComponent />
          </RouteTitle>
        }
      />
      <Route
        path="/search"
        element={
          <RouteTitle title="Search">
            <SearchComponent />
          </RouteTitle>
        }
      />
      <Route
        path="/download"
        element={
          <RouteTitle title="Download">
            <DownloadComponent />
          </RouteTitle>
        }
      />
      <Route
        path="/manual"
        element={
          <RouteTitle title="Manual">
            <ManualAddComponent />
          </RouteTitle>
        }
      />
      <Route
        path="/stats"
        element={
          <RouteTitle title="Stats">
            <StatsComponent />
          </RouteTitle>
        }
      />
      <Route
        path="/diagnostics"
        element={
          <RouteTitle title="Diagnostics">
            <DiagnosticsComponent />
          </RouteTitle>
        }
      />
      <Route
        path="/storybook"
        element={
          <RouteTitle title="Storybook">
            <Storybook />
          </RouteTitle>
        }
      />
      <Route
        path="/monitor/delete/:id"
        element={
          <RouteTitle title="Monitor">
            <MonitorDeleteComponent />
          </RouteTitle>
        }
      />
      <Route
        path="/monitor/add/:tmdb_id"
        element={
          <RouteTitle title="Monitor">
            <MonitorAddComponent />
          </RouteTitle>
        }
      />
      <Route
        path="/monitor"
        element={
          <RouteTitle title="Monitor">
            <MonitorComponent />
          </RouteTitle>
        }
      />
      <Route
        path="/"
        element={
          <RouteTitle title="Media">
            <IndexComponent />
          </RouteTitle>
        }
      />
    </Routes>
  );
}
export { ParentComponent };
