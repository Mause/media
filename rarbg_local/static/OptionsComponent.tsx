import _ from 'lodash';
import qs from 'qs';
import React, { Component, useState, useEffect } from 'react';
import { Season } from './SeasonSelectComponent';
import { load, subscribe, useLoad } from './utils';
import { Torrents } from './streaming';
import { withRouter, RouteComponentProps, Link } from 'react-router-dom';

type OptionsProps = { type: 'movie' | 'series' } & RouteComponentProps<{ tmdb_id: string, season?: string, episode?: string }>;


function getHash(magnet: string) {
  const u = new URL(magnet);
  return _.last(u.searchParams.get('xt')!.split(':'))
}

interface ITorrent {
  title: string;
  seeders: number;
  download: string;
}

interface ItemInfo {
  imdb_id: string;
  title: string;
}

function DisplayTorrent({ torrent, torrents, itemInfo, type }: { torrent: ITorrent, torrents?: Torrents, itemInfo?: ItemInfo, type: string }) {
  let url;
  if (itemInfo) {
    url = {
      pathname: `/download`, state: {
        type,
        imdb_id: itemInfo.imdb_id,
        magnet: torrent.download,
        titles: itemInfo.title,
      }
    };
  }
  return (
    <span>
      {url ? <Link to={url}>{torrent.title}</Link> : torrent.title}
      &nbsp;
      <small>{torrent.seeders}</small>
      &nbsp;
      {torrents && getHash(torrent.download)! in torrents ? <small>downloaded</small> : ''}
    </span>
  );
}

const ranking = [
  'Movies/XVID',
  'Movies/x264',
  'Movies/x264/720',
  'Movies/XVID/720',
  'Movies/BD Remux',
  'Movies/Full BD',
  'Movies/x264/1080',
  'Movies/x264/4k',
  'Movies/x265/4k',
  'Movies/x264/3D',
  'Movs/x265/4k/HDR',

  'TV Episodes',
  'TV HD Episodes',
  'TV UHD Episodes',
  'TV-UHD-episodes',
];

function remove(bit: string): string {
  if (bit.startsWith('Mov')) {
    const parts = bit.split('/');
    return parts.slice(1).join('/');
  }
  return bit;
}

class _OptionsComponent extends Component<OptionsProps, {
  results: ITorrent[];
  loading: boolean;
  itemInfo?: ItemInfo;
}> {
  constructor(props: OptionsProps) {
    super(props);
    this.state = { results: [], loading: true, itemInfo: undefined };
  }
  public render() {
    return <Pure
      itemInfo={this.state.itemInfo}
      type={this.props.type}
      episode={this.props.match.params.episode}
      season={this.props.match.params.season}
      tmdb_id={this.props.match.params.tmdb_id}
    />;
  }
  componentDidMount() {
    const { tmdb_id, season, episode } = this.props.match.params;
    const prom = load<ItemInfo>(`${this.props.type == 'series' ? 'tv' : 'movie'}/${tmdb_id}`);
    if (this.props.type === 'series') {
      Promise.all([
        load<Season>(`tv/${tmdb_id}/season/${season}`),
        prom
      ]).then(([season, { imdb_id }]) => {
        this.setState({ itemInfo: { title: season.episodes[parseInt(episode!, 10)].name, imdb_id } });
      });
    }
    else {
      prom.then(itemInfo => this.setState({ itemInfo }));
    }
  }
}

function Pure(props: { itemInfo?: ItemInfo, type: string, tmdb_id: string, season?: string, episode?: string }) {
  const torrents = useLoad<Torrents>('torrents');
  const { type } = props;
  const { items: results, loading } = useSubscribe<ITorrent>(`/stream/${type}/${props.tmdb_id}?` + qs.stringify({ season: props.season, episode: props.episode }));
  const dt = (result: ITorrent) => <DisplayTorrent itemInfo={props.itemInfo} type={type} torrents={torrents} torrent={result} />;
  const grouped = _.groupBy(results, 'category');
  const auto = _.maxBy(grouped['Movies/x264/1080'] || grouped['TV HD Episodes'] || [], 'seeders');
  const bits = _.sortBy(_.toPairs(grouped), ([category]) => -ranking.indexOf(category)).map(([category, results]) => (<div key={category}>
    <h3>{remove(category)}</h3>
    {_.sortBy(results, i => -i.seeders).map(result => (<li key={result.title}>{dt(result)}</li>))}
  </div>));
  return (<div>
    {loading ? <i className="fas fa-spinner fa-spin fa-xs"></i> : ''}
    {bits.length || loading ?
      <div>
        <p>
          Auto selection: {auto ? dt(auto) : 'None'}
        </p>
        <ul>{bits}</ul>
      </div> :
      'No results'}
  </div>);
}

function useSubscribe<T>(url: string) {
  const [subscription, setSubscription] = useState<{ items: T[], loading: boolean }>({ loading: true, items: [] });

  useEffect(() => {
    const items: T[] = [];
    subscribe(url, data => {
      items.push(data);
      setSubscription({
        items,
        loading: true,
      })
    }, () => setSubscription({ loading: false, items }));
  }, [url]);

  return subscription;
}


const OptionsComponent = withRouter(_OptionsComponent);
export { OptionsComponent };
