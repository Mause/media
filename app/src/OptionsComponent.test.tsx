import React from 'react';
import { render, act, wait } from '@testing-library/react';
import { OptionsComponent } from './OptionsComponent';
import { swrConfig } from './streaming';
import { MemoryRouter, Route } from 'react-router-dom';

import EventSource from 'eventsource';

let sources = [];
class ES {
  constructor() {
    sources.push(this);
    console.warn(this);
  }
  addEventListener(name, ls) {
    this.ls = ls;
  }
  close() {}
}

Object.defineProperty(window, 'EventSource', {
  value: ES, //EventSource,
});

test('OptionsComponent', async () => {
  await act(async () => {
    const el = render(
      <MemoryRouter>
        <Route p>
          <OptionsComponent />
        </Route>
      </MemoryRouter>,
    );
    expect(el.container).toMatchSnapshot();

    await wait(1);

    sources[0].ls({ data: '[{}]' });
  });
});
