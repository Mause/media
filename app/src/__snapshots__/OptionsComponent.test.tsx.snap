// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsComponent 1`] = `
<div>
  <div>
    <i
      class="fas fa-spinner fa-spin fa-xs"
    />
    <div>
      <p>
        Auto selection: 
        None
      </p>
      <ul />
    </div>
  </div>
</div>
`;
